doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,955 @@
+One of the first questions we should be considering is how we represent programs
+in such a way that we can easily analyze and reason them.
+
+As it turns out, the best way to do this is to use a variety of
+different representations, each representing a different level of
+abstraction. These representations are designed in such a way that for
+each analysis we want to perform, there is an appropriate
+representation, and it is easy to go from representations that are
+close to the source code to representations that focus on specific
+semantic aspects of the program.
+
+The representations that the CPROVER framework uses mirror those
+used in modern compilers such as LLVM and gcc. I will point out those
+places where the CPROVER framework does things differently, attempting to give
+rationales wherever possible.
+
+One in-depth resource for most of this section is the classic
+(compiler construction text book)[https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools]
+by Aho, Lam, Sethi and Ullman.
+
+To illustrate the different concepts, we will consider a small example
+program. While the program is in C, the general ideas apply to other
+languages as well - see later sections of this manual to understand how
+the specific features of those languages are handled. Our running
+example will be a program that calculates factorials.
+
+```C
+/* factorial.c
+ *
+ * For simplicity's sake, we just give the forward
+ * declarations of atoi and printf.
+ */
+int atoi(const char *);
+int printf(const char *, ...);
+
+unsigned long factorial(unsigned n) {
+  unsigned long fac = 1;
+  for (unsigned int i = 1; i <= n; i++) {
+    fac *= i;
+  }
+  return fac;
+}
+
+/* Error handling elided - this is just for illustration. */
+int main(int argc, const char **argv) {
+  unsigned n = atoi(argv[1]);
+  printf("%u! = %lu\n", n, factorial(n));
+  return 0;
+}
+```
+
+The question of this first section is: How do we represent this program
+in memory so that we can do something useful with it?
+
+One possibility would be to just store the program as a string, but this
+is clearly impractical: even finding whether there is an assignment to a
+specific variable would require significant parsing effort. For this
+reason, the first step is to parse the program text into a more abstract
+representation.
+
+\subsection AST_section AST
+
+The first step in representing a program in memory is to parse the
+program, at the same time checking for syntax errors, and store the
+parsing result in memory.
+
+The key data structure that stores the result of this step is known as an
+**Abstract Syntax Tree**, or **AST** for short
+(cf. [Wikipedia](https://en.wikipedia.org/wiki/Abstract_syntax_tree)).
+ASTs are still relatively
+close to the source code, and represent the structure of the source code
+while abstracting away from irrelevant details, e.g., dropping
```

Comment:
Don't use the term "irrelevant". You don't know what the reader may or may not consider relevant.